corehq/util/decorators.py

```diff
@@ -62,3 +64,65 @@ def __enter__(self):
 
     def __exit__(self, exc_type, exc_val, exc_tb):
         pass
+
+
+class CouldNotAqcuireLock(Exception):
+    pass
+
+
+def _get_unique_key(format_str, fn, *args, **kwargs):
+    # Sorry this is so magic
+    varnames = fn.func_code.co_varnames
+    kwargs.update(dict(zip(varnames, args)))
+    return ("{}-" + format_str).format(fn.__name__, **kwargs)
+
+
+def locking_task(unique_key, default_retry_delay=30, timeout=5*60, max_retries=3):
+    """
+    Define a locking celery task to prevent race conditions.
+
+    :param unique_key: string used to lock the task.  There will be one lock
+        per unique value.  You may use any arguments that will be passed to the
+        function.  See example.
+    :param default_retry_delay: seconds to wait before retrying if a lock is
+        encountered
+    :param timeout: timeout on the lock (in seconds).  Normally the lock should
+        be released when the task completes, but you should also define a
+        timeout in case something goes wrong.  This must be greater than the
+        maximum length of the task.
+
+    Usage:
+        @locking_task("{user.username}-{from}", default_retry_delay=2)
+        def greet(user, from="Dimagi"):
+            ...
+
+        greet.delay(joeshmoe)
+        # Locking key used would be "greet-joeshmoe@test.commcarehq.org-Dimagi"
+    """
```

Comment:
This is the interface I wanted for a one-at-a-time task, so I made it.  It makes sense to me, but happy to hear suggestions.  There is a lot of boilerplate around this, so I figured a nice wrapper would make the code a lot more comprehensible.  Plus, I imagine this will be generally useful.